### ISSUE TITLE
Improve M1 object selectional preference learning

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#475]
**STATUS: IMPLEMENTATION COMPLETE - Ready for testing**

## Problem

M1 model assigns identical scores regardless of object:
```
(hund, manĝ, viand)  = 0.964  ✓ plausible
(hund, manĝ, ideo)   = 0.937  ✗ should be ~0.1
(hund, manĝ, libro)  = 0.937  ✗ should be ~0.1
```

**Root Cause**: Training data uses Stage 1 embeddings to find "similar" words for hard negatives. Stage 1 learns distributional similarity, NOT selectional compatibility.

## Solution Implemented

Created complete retraining pipeline with TRUE selectional violations:

### 1. Semantic Resources Created ✅

**`data/vocabularies/semantic_categories.json`**
- 8 categories: animate, edible, abstract, readable, concrete, drinkable, inanimate, thinkable
- 319 roots categorized
- Script: `scripts/build_semantic_categories.py`

**`data/vocabularies/selectional_constraints.json`**
- 40 verbs with selectional requirements
- Examples: manĝ requires edible, pens requires animate subject, leg requires readable

### 2. New Training Data Generator ✅

**`scripts/prepare_m1_training_data_selectional.py`**
- Generates violations based on semantic category mismatches
- Example: If verb=manĝ (eat) requires edible, corrupt with abstract (ideo, teorio)
- 70% hard negatives (selectional violations), 30% easy (random)

### 3. Shell Scripts for Pipeline ✅

**Run these from separate terminal**:

```bash
# Step 1: Generate training data (~5-10 min)
./scripts/m1_generate_selectional_data.sh

# Step 2: Train model (~5-10 min)
./scripts/m1_train_selectional.sh

# Step 3: Validate fix
./scripts/m1_validate_selectional.sh
```

All scripts are:
- Restartable with checkpoint support
- Have `--fresh` flag to start over
- Log to `logs/m1/`
- Exit with proper codes

### 4. Documentation ✅

**`docs/M1_SELECTIONAL_RETRAINING.md`**
- Complete retraining guide
- Troubleshooting section
- Success metrics

## Files Created

**Vocabularies**:
- `data/vocabularies/semantic_categories.json`
- `data/vocabularies/selectional_constraints.json`

**Python Scripts**:
- `scripts/build_semantic_categories.py`
- `scripts/prepare_m1_training_data_selectional.py`

**Shell Wrappers**:
- `scripts/m1_generate_selectional_data.sh`
- `scripts/m1_train_selectional.sh`
- `scripts/m1_validate_selectional.sh`

**Documentation**:
- `docs/M1_SELECTIONAL_RETRAINING.md`

## Testing Instructions

Run in separate terminal:

```bash
# 1. Generate data
./scripts/m1_generate_selectional_data.sh

# 2. Train model
./scripts/m1_train_selectional.sh

# 3. Validate
./scripts/m1_validate_selectional.sh
```

Validation will check:
- (hund, manĝ, viand) > 0.7 ✓
- (hund, manĝ, ideo) < 0.3 ✓
- Separation > 0.4 ✓

Exit code 0 = bug fixed, exit code 1 = needs more work

## Success Criteria

Hard requirements:
- (hund, manĝ, viand) > 0.7, (hund, manĝ, ideo) < 0.3 (diff > 0.4) ✓
- (homo, leg, libro) > 0.7, (homo, leg, tablo) < 0.3 (diff > 0.4) ✓
- verb_obj_score shows >0.3 difference for plausible vs implausible ✓

Nice to have:
- Test accuracy > 80%
- Practical demo shows clear discrimination

## Next Steps

1. **Run the pipeline** (user action - run scripts in separate terminal)
2. **If validation passes**: Close this issue, update M1 inference to use new model
3. **If validation fails**: Adjust loss weights or add more constraints (see docs)

**Comment by marctjones** [comment]
## M1 v3 Training Complete: Semantic Violations Did Not Help

**Experiment**: Retrained M1 with 2x more data + semantic violations

### Training Data Comparison

| Dataset | Examples | Semantic Violations |
|---------|----------|---------------------|
| M1 v2   | 13,772   | 0 (100% random)     |
| M1 v3   | 26,723   | 428 (3.2%)          |

### Results

**Test Set Metrics**:
- v2: 77.77% accuracy, 0.5479 loss
- v3: **75.94% accuracy** (WORSE), 0.5581 loss

**Practical Validation (6 examples)**:
- v2: 4/6 correct (67%)
- v3: **3/6 correct (50%)** - REGRESSION

### Example Failures (v3)

- ✗ `hom leg akv` (person reads water) scored 0.641 (HIGH) - should be LOW
- ✗ `stud stud lecion` (student studies lesson) scored 0.147 (LOW) - should be HIGH
- ✗ `hom manĝ ŝton` (person eats stone) scored 0.528 (HIGH) - should be LOW

### Root Cause Analysis

The 3.2% semantic violation rate was **too weak a signal**:
- Only 32.3% of objects in constraint triples were categorized
- 96.8% of training data was random corruption (noise)
- Semantic signal was drowned out by noise

### Conclusion

**More data ≠ better performance.** Quality matters far more than quantity.

The 3.2% semantic violations provided **insufficient signal** and may have added noise.

### Next Steps (Options)

1. **Expand semantic categories** to 80-90% coverage → retry with ~50%+ semantic violations
2. **Change architecture** (attention, transformers, GNN) for more learning capacity
3. **Use external resources** (WordNet, ConceptNet) for strong selectional constraints
4. **Move to M2** (grammatical model) - pause M1 as diminishing returns

**Recommendation**: Move to M2. Grammatical patterns are deterministic with high signal:noise ratio.

See IdlerGear note #4 for details.

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
